engine/src/sellst.cpp

```diff
@@ -192,149 +176,120 @@ void MCSellist::add(MCObject *objptr, bool p_sendmessage)
-		uint4 i;
-		MCSelnode *newobjects = NULL;
-		for (i = 0 ; i < nitems ; i++)
-		{
-			optr = (MCSelnode *)items[i].data;
-			optr->remove(objects);
-			optr->appendto(newobjects);
-		}
-		objects = newobjects;
-	}
+
+    if (m_objects.size() < 2)
+        return;
+
+    /* Create a list of object handles with their layer numbers. We do
+     * this because computing the layer number is relatively
+     * expensive. */
+    struct SortNode
+    {
+        SortNode() = default;
+        SortNode(MCObject* o, uint2 n) : m_object(o), m_num(n) {}
+        MCObject *m_object {};
+        uint2 m_num {};
+        bool operator<(const SortNode other) const { return m_num < other.m_num; }
+    };
+
+    /* UNCHECKED */ std::vector<SortNode> t_items {m_objects.size()};
+    MCCard *t_card = m_objects.front()->getcard();
+
+    std::transform(m_objects.begin(), m_objects.end(),
+                   t_items.begin(),
+                   [&](const MCSelnode& t_node) -> SortNode {
+                       uint2 t_num {};
+                       t_card->count(CT_LAYER, CT_UNDEFINED,
+                                     t_node, t_num, True);
+                       return {t_node.Get(), t_num};
+                   });
+    m_objects.clear();
+
+    /* Sort the object handles by layer number */
+    std::sort(t_items.begin(), t_items.end());
+
+    /* Rebuild the selection in the appropriate order */
+    for (const auto& t_sorted : t_items)
+        m_objects.emplace_back(MCObjectHandle(t_sorted.m_object));
 }
 
 uint32_t MCSellist::count()
 {
-	uint32_t t_count = 0;
-	MCSelnode *t_node = objects;
-	if (t_node != nil)
-	{
-		do
-		{
-            // Only count non-zombie objects
-            if (t_node->m_ref)
-                t_count++;
-            
-			t_node = t_node->next();
-		} while (t_node != objects);
-	}
-
-	return t_count;
+    return std::count_if(m_objects.begin(), m_objects.end(),
+                         [](const MCSelnode& t_node) {
+                             return static_cast<bool>(t_node);
+                         });
 }
 
 MCControl *MCSellist::clone(MCObject *target)
 {
-    // Remove any dead objects before trying to clone them
-    Clean();
-    
-    MCObjectHandle *t_selobj_handles = nil;
-	uint32_t t_selobj_count;
-	t_selobj_count = count();
-
-	/* UNCHECKED */ MCMemoryNewArrayInit(t_selobj_count, t_selobj_handles);
-	sort();
-
-	MCSelnode *t_node = objects;
-	for (uint32_t i = 0; i < t_selobj_count; i++)
-	{
-		t_selobj_handles[i] = t_node->m_ref->GetHandle();
-		t_node = t_node->next();
-	}
+    /* Remove any dead objects before trying to clone them & sort
+     * what's left */
+    sort();
+
+    /* Work through the selection, replacing each selected object's
+     * entry with its clone. */
+    MCControlHandle t_newtarget;
+    std::transform(
+        m_objects.begin(), m_objects.end(), m_objects.begin(),
+        [&](const MCSelnode& t_node) {
```

Comment:
Comparing with the lambda above, this one has no return type specification - is this because we are making the return type explicit due to the object constructor here, or is there a less obvious rule at play?